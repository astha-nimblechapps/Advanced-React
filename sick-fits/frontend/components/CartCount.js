import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { TransitionGroup, CSSTransition } from "react-transition-group";

const Dot = styled.div`
  background: ${props => props.theme.black};
  color: white;
  border-radius: 50%;
  padding: 0.5rem;
  line-height: 2.5rem;
  min-width: 3.5rem;
  margin-left: 1rem;
  font-weight: 100;
  font-feature-settings: "tnum";
  font-variant-numeric: tabular-nums;
  text-align: center;
`;

const AnimatedStyles = styled.span`
  position: relative;
  .count {
    display: block;
    position: relative;
    transition: all 0.4s;
    backface-visibility: hidden;
  }
  .count-enter {
    transform: scale(4) rotateX(0.5turn);
  }
  .count-enter-active {
    transform: rotateX(0);
  }
  .count-exit {
    top: 0;
    position: absolute;
    transform: rotateX(0);
  }
  .count-exit-active {
    transform: scale(4) rotateX(0.5turn);
  }
`;

const CartCount = props => (
  <AnimatedStyles>
    <TransitionGroup>
      <CSSTransition
        unmountOnExit
        className="count"
        classNames="count"
        key={props.count}
        timeout={{ enter: 400, exit: 400 }}
      >
        <Dot>{props.count}</Dot>
      </CSSTransition>
    </TransitionGroup>
  </AnimatedStyles>
);

export default CartCount;
