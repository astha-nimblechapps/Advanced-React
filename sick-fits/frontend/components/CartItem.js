import React from "react";
import formatMoney from "../lib/formatMoney";
import styled from "styled-components";
import PropTypes from "prop-types";
import RemoveFromCart from "./RemoveFromCart";
import { Typography, Label } from "./../lib/exim-component";

const CartItemStyle = styled.div`
  padding: 1rem 0;
  border-bottom: 1px solid ${props => props.theme.lightgrey};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 10px;
  }
  h3,
  p {
    margin: 0px;
  }
`;

const CartItem = props => {
  if (!props.cartItem.item)
    return (
      <CartItemStyle>
        <p>This item has been removed</p>{" "}
        <RemoveFromCart id={props.cartItem.id} />
      </CartItemStyle>
    );
  return (
    <CartItemStyle>
      <img
        src={props.cartItem.item.image}
        alt={props.cartItem.item.title}
        width="100"
      />
      <div className="cart-item-details">
        <Typography useFor="subtitle">{props.cartItem.item.title}</Typography>
        <p>
          {props.cartItem.color}
          {props.cartItem.size}
        </p>
        <p>
          <Label color="green">
            {formatMoney(props.cartItem.item.price * props.cartItem.quantity)}
          </Label>
          {" - "}
          <Label color="lightNeutral">
            <em>
              {props.cartItem.quantity} &times;{" "}
              {formatMoney(props.cartItem.item.price)}
            </em>
          </Label>
        </p>
      </div>
      <RemoveFromCart id={props.cartItem.id} />
    </CartItemStyle>
  );
};

CartItem.propTypes = {
  cartItem: PropTypes.object.isRequired
};

export default CartItem;
