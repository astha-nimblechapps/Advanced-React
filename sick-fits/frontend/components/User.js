import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';

const LOGGED_USER = gql`
  query {
    me {
      id
      email
      name
      permissions
    }
  }
`;

const User = props => (
  <Query {...props} query={LOGGED_USER}>
    {payload => props.children(payload)}
  </Query>
);

User.PropTypes = {
  children: PropTypes.func.isRequired,
};

export default User;
export { LOGGED_USER };
