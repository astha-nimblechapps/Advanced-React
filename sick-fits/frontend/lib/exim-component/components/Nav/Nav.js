"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var classnames_1 = require("classnames");
var NavItem_1 = require("./NavItem");
exports.NavItem = NavItem_1.default;
var Nav = /** @class */ (function (_super) {
    __extends(Nav, _super);
    function Nav(props) {
        var _this = _super.call(this, props) || this;
        var key = _this.props.selectedKey ? _this.props.selectedKey : null;
        _this.state = {
            selectedKey: key
        };
        return _this;
    }
    Nav.prototype.onSelect = function (event, key) {
        this.setState({
            selectedKey: key
        });
        if (this.props.onSelect)
            this.props.onSelect(event, key);
    };
    Nav.prototype.getClasses = function () {
        var _a = this.props, customClass = _a.customClass, className = _a.className;
        return classnames_1.default(customClass, className);
    };
    Nav.prototype.getChildren = function () {
        var _this = this;
        var _a = this.props, children = _a.children, customClass = _a.customClass;
        return React.Children.map(children, function (child, index) {
            var key = child.key ? child.key : (index + 1);
            var originalOnClick = child.props.onClick;
            child = React.cloneElement(child, {
                customClass: customClass + '-' + 'item',
                key: key,
                selected: (_this.state.selectedKey == key) ? true : false,
                onClick: function (event) {
                    if (!child.props.disabled) {
                        _this.onSelect(event, key);
                        if (originalOnClick) {
                            originalOnClick.apply(child, event);
                        }
                    }
                }
            });
            return child;
        });
    };
    Nav.prototype.render = function () {
        var boundChildren = this.getChildren();
        var classes = this.getClasses();
        return (React.createElement("ul", { className: classes }, boundChildren));
    };
    Nav.defaultProps = {
        customClass: 'nav',
        disabled: false
    };
    return Nav;
}(React.Component));
exports.Nav = Nav;
exports.default = Nav;
//# sourceMappingURL=Nav.js.map