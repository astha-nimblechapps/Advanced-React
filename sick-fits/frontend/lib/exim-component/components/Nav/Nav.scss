.left_sidebar {
  width: 220px !important;
  padding: 40px 30px;
  float: left;
  background: #F7F8F9;

  .nav {
    padding: 0;
    margin: 0;
    list-style: none;

    .nav-item {
      margin: 5px 0;
      display: block;

      span {
        cursor: pointer;
        position: relative;
        display: inline-block;
        border-radius: 0;
        padding: 5px 0;
        color: #626B76;
        padding-left: 30px;
        font-size: 16px;
        font-family: 'DINRoundPro' ,sans-serif;
        text-decoration: none;

        i {
          font-size: 22px;
          position: absolute;
          left: 0;
          top: 3px;
        }
      }
    }

    .nav-item.nav-item-selected span {
      color: #0667D0;
    }

    .nav-item span:hover,
    .nav-item.nav-item-selected span:hover {
      color: #0667D0;
    }

    .nav-item.nav-item-disabled span {
      cursor: not-allowed;
    }

    .nav-item.nav-item-disabled span:hover {
      color: #626B76;
    }
  }
}
