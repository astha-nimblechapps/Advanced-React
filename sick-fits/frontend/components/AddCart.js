import React from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import styled from "styled-components";
import User, { LOGGED_USER } from "./User";

const ADDCART = styled.button`
  width: 30%;
  /* border: 1px solid; */
  text-align: center;
  font-family: none;
  border: none;
`;

const ADD_TO_CART = gql`
  mutation ADD_TO_CART($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
    }
  }
`;

const ADD_TO_TEMP_CART = gql`
  mutation ADD_TO_TEMP_CART(
    $id: ID!
    $token: String!
    $color: String
    $size: String
  ) {
    addToTempCart(id: $id, token: $token, color: $color, size: $size) {
      id
      quantity
      color
      size
    }
  }
`;
/**
 * e.preventDefault();
    const res = await updateItemMutation({
      variables: {
        id: this.props.id,
        ...this.state
      }
    });
    console.log("Updated..!!!");
 */

class AddCart extends React.Component {
  addTempCart = async (e, addTempCart) => {
    e.preventDefault();
    if (this.props.color === '' && this.props.size === '' && this.props.color === 'none' && this.props.size === 'none') {
      alert("Please Select Color and Size both")
    } else {
      if (localStorage.getItem("randomId")) {
        console.log("true");
        await addTempCart({
          variables: {
            id: this.props.id,
            token: localStorage.getItem("randomId"),
            color: this.props.color,
            size: this.props.size
          }
        });
      } else {
        console.log("false");
        const min = 1;
        const max = 100;
        const rand = min + Math.random() * (max - min);
        // console.log(rand)
        this.setState({ token: rand });
        localStorage.setItem("randomId", rand);
        // console.log(JSON.stringify(localStorage.getItem("randomId")))
        await addTempCart({
          variables: {
            id: this.props.id,
            token: localStorage.getItem("randomId"),
            color: this.props.color,
            size: this.props.size
          }
        });
      }
    }
  };

  render() {
    const id = this.props.id;

    return (
      <User>
        {({ data: { me } }) => {
          if (me)
            return (
              <Mutation
                mutation={ADD_TO_CART}
                variables={{
                  id
                }}
                refetchQueries={[{ query: LOGGED_USER }]}
              >
                {(addToCart, { loading }) => (
                  <ADDCART disabled={loading} onClick={addToCart}>
                    Add{loading && "ing"}To Cart
                  </ADDCART>
                )}
              </Mutation>
            );
          else {
            return (
              <Mutation mutation={ADD_TO_TEMP_CART}>
                {(addToTempCart, { loading }) => (
                  <ADDCART
                    disabled={loading}
                    onClick={e => this.addTempCart(e, addToTempCart)}
                  >
                    Add{loading && "ing"}To Cart
                  </ADDCART>
                )}
              </Mutation>
            );
          }
        }}
      </User>
    );
  }
}

export default AddCart;
export { ADD_TO_CART, ADDCART };
