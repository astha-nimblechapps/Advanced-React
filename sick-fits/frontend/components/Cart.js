import React from "react";
import { Query, Mutation } from "react-apollo";
import gql from "graphql-tag";
import { adopt } from "react-adopt";
import CloseButton from "./styles/CloseButton";
import Supreme from "./styles/Supreme";
import SickButton from "./styles/SickButton";
import CartStyles from "./styles/CartStyles";
import User from "./User";
import CartItem from "./CartItem";
import calcTotalPrice from "../lib/calcTotalPrice";
import formatMoney from "../lib/formatMoney";
import Payment from "./Payment";
import Router from 'next/router';

const LOCAL_STATE = gql`
  query {
    cartOpen @client
  }
`;

const TOOGLE_CART = gql`
  mutation {
    toogleCart @client
  }
`;

const TEMP_DATA = gql`
  query($token: String) {
    tempCartItem(token: $token) {
      id
      color
      size
      quantity
      item {
        image
        id
        price
        description
        title
      }
    }
  }
`;



class Cart extends React.Component {
  
  
  constructor(props) {
    super(props);
    this.state = {
      tokenId: ""
    }
  }


  componentDidMount() {
    if(localStorage.getItem("randomId")){
      this.setState({
        tokenId: localStorage.getItem("randomId").toString()
      })
    }
    // this.setState({
    //   randomUser: localStorage.getItem("randomId")
    // });
  }

  goToLogin = async(toogleCart) =>{
    await toogleCart();
    Router.push({
      pathname:'/signin'
    })
  }
  
  render() {
    const Composed = adopt({
      user: ({ render }) => <User>{render}</User>,
      toogleCart: ({ render }) => (
        <Mutation mutation={TOOGLE_CART}>{render}</Mutation>
      ),
      localState: ({ render }) => <Query query={LOCAL_STATE}>{render}</Query>,
      tempCart: ({ render }) => <Query query={TEMP_DATA} variables={{ token : this.state.tokenId}}>{render}</Query>
    });
    return (
      <Composed>
        {({ user, toogleCart, localState, tempCart }) => {
          const me = user.data.me;
          const cart = tempCart.data.tempCartItem;
          console.log(cart)
          if (!me){ 
            if(!cart) return <p> No item</p>
            return ( 
            <CartStyles open={localState.data.cartOpen}>
            <header>
              <CloseButton onClick={toogleCart} title="close">
                &times;
              </CloseButton>
              <Supreme>Your Cart</Supreme>
              <p>
                You have {cart.length} itme
                {cart.length === 1 ? "" : "s"} in your cart
              </p>
            </header>
            <ul>
              {cart.map(cartItem => (
              //  console.log(cartItem)
                <CartItem key={cartItem.id} cartItem={cartItem} />
              ))}
            </ul>
            <footer>
              <p>{formatMoney(calcTotalPrice(cart))}</p>
           
              {
                cart.length ? <Payment cart={cart}>
                <SickButton onClick={() => this.goToLogin(toogleCart)}>Checkout</SickButton>
              </Payment> : null
              }
              {/* {cart.length && (
                <Payment>
                  <SickButton>Checkout</SickButton>
                </Payment>
              )} */}
            </footer>
            </CartStyles>
          )
        }
          return (
            <CartStyles open={localState.data.cartOpen}>
              <header>
                <CloseButton onClick={toogleCart} title="close">
                  &times;
                </CloseButton>
                <Supreme>Your Cart</Supreme>
                <p>
                  You have {me.cart.length} itme
                  {me.cart.length === 1 ? "" : "s"} in your cart
                </p>
              </header>
              <ul>
                {me.cart.map(cartItem => (
                  <CartItem key={cartItem.id} cartItem={cartItem} />
                ))}
              </ul>
              <footer>
                <p>{formatMoney(calcTotalPrice(me.cart))}</p>
                {me.cart.length && (
                  <Payment>
                    <SickButton>Checkout</SickButton>
                  </Payment>
                )}
              </footer>
            </CartStyles>
          );
        }}
      </Composed>
    );
  }
}

export default Cart;
export { TOOGLE_CART, LOCAL_STATE, TEMP_DATA };
