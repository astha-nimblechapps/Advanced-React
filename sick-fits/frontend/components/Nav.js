import Link from "next/link";
import { Mutation } from "react-apollo";
import styled from "styled-components";
import NavStyles from "./styles/NavStyles";
import NProgress from "nprogress";
import Router from "next/router";
import User from "./User";
import SignOut from "./SignOut";
import { TOOGLE_CART } from "./Cart";
import CartCount from "./CartCount";

Router.onRouteChangeStart = () => {
  NProgress.start();
};
Router.onRouteChangeComplete = () => {
  NProgress.done();
};

Router.onRouteChangeError = () => {
  NProgress.done();
};

const NavDiv = styled.div`
  margin-left: 10px;
`;

const Nav = () => (
  <User>
    {({ data: { me } }) => (
      <NavStyles data-test="nav">
        {me && (
          <Link href="/">
            <a> {me.name}</a>
          </Link>
        )}
        <Link href="/items">
          <a> Item </a>
        </Link>
        {me && (
          <>
            <Link href="/sell">
              <a> Sell </a>
            </Link>
            <Link href="/orders">
              <a> My Orders </a>
            </Link>
            <SignOut />
            <Mutation mutation={TOOGLE_CART}>
              {toogleCart => (
                <button onClick={toogleCart}>
                  Cart
                  <CartCount
                    count={me.cart.reduce(
                      (tally, cartItem) => tally + cartItem.quantity,
                      0
                    )}
                  />
                </button>
              )}
            </Mutation>
          </>
        )}
        {!me && (
          <Link href="/singup">
            <a> Signup </a>
          </Link>
        )}
      </NavStyles>
    )}
  </User>
);

export default Nav;
