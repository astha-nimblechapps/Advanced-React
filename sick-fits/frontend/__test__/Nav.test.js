import { mount } from "enzyme";
import toJSON from 'enzyme-to-json';
import wait from "waait";
import Nav from "../components/Nav";
import { LOGGED_USER } from "../components/User";
import { MockedProvider } from "react-apollo/test-utils";
import { fakeUser, fakeCartItem } from "../lib/testUtils";

const notSignIn = [
  {
    request: {
      query: LOGGED_USER
    },
    result: {
      data: { me: null }
    }
  }
];

const signIn = [
  {
    request: {
      query: LOGGED_USER
    },
    result: {
      data: { me: fakeUser() }
    }
  }
];

const signedInMocksWithCartItems = [
    {
      request: { query: LOGGED_USER },
      result: {
        data: {
          me: {
            ...fakeUser(),
            cart: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
          },
        },
      },
    },
  ];

describe("<Nav/>", () => {
  it("render signed out nav ", async () => {
    const wrapper = mount(
      <MockedProvider mocks={notSignIn}>
        <Nav />
      </MockedProvider>
    );

    await wait();
    wrapper.update();
   // console.log(wrapper.debug());
    expect(toJSON(wrapper.find('[data-test="nav"]'))).toMatchSnapshot();
  });

  it("render signed in nav",async () =>{
    const wrapper = mount(
        <MockedProvider mocks={signIn}>
          <Nav />
        </MockedProvider>
      );

      await wait();
    wrapper.update();
    //console.log(wrapper.debug());
    const nav = wrapper.find('ul[data-test="nav"]');
    expect(nav.children().length).toBe(6);
    expect(nav.text()).toContain('Sign Out');
  });

  it('renders the amount of items in the cart', async () => {
    const wrapper = mount(
      <MockedProvider mocks={signedInMocksWithCartItems}>
        <Nav />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
   
    const nav = wrapper.find('[data-test="nav"]');
    //console.log(nav.debug());
    const count = nav.find('div.count');
    expect(toJSON(count)).toMatchSnapshot();
  });
});
