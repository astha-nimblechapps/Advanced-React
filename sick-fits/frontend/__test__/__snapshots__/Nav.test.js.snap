// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav/> render signed out nav  1`] = `
Array [
  <NavStyles
    data-test="nav"
  >
    <ul
      className="NavStyles-sc-11c0d2g-0 fWyNOv"
      data-test="nav"
    >
      <Link
        href="/items"
      >
        <a
          href="/items"
          onClick={[Function]}
        >
           Item 
        </a>
      </Link>
      <Link
        href="/singup"
      >
        <a
          href="/singup"
          onClick={[Function]}
        >
           Signup 
        </a>
      </Link>
    </ul>
  </NavStyles>,
  <ul
    className="NavStyles-sc-11c0d2g-0 fWyNOv"
    data-test="nav"
  >
    <Link
      href="/items"
    >
      <a
        href="/items"
        onClick={[Function]}
      >
         Item 
      </a>
    </Link>
    <Link
      href="/singup"
    >
      <a
        href="/singup"
        onClick={[Function]}
      >
         Signup 
      </a>
    </Link>
  </ul>,
]
`;

exports[`<Nav/> renders the amount of items in the cart 1`] = `
<div
  className="count CartCount__Dot-xxvp4g-0 iAflQU"
>
  9
</div>
`;
